### PR TITLE
Update scanline functions and get_pr/lr_pts function in series class.

### PR DESCRIPTION
**Pull request by linwang9926**
- Update scanline functions
- Update get_pr/lr_pts function in series class.
- Update stem width function

### CODE REVIEW

File: sleap_roots/graphpipeline.py [hunk 1]
```diff
@@ -238,16 +234,19 @@ def get_traits_value_frame(
...
         fn_outputs = fn(*[input_trait for input_trait in inputs])
         if type(fn_outputs) == tuple:
             fn_outputs = np.array(fn_outputs).reshape((1, -1))
-        if type(fn_outputs) == np.ndarray and fn_outputs.shape == (fn_outputs.size,):
-            fn_outputs = fn_outputs.reshape((1, -1))
+        if (
+            type(fn_outputs) == np.float64
+            or type(fn_outputs) == float
+            or type(fn_outputs) == np.int64
+            or type(fn_outputs) == int
```
Comment (talmo) [inline sleap_roots/graphpipeline.py:245]:
```suggestion
            isinstance(fn_outputs, (np.floating, float)) 
            or isinstance(fn_outputs, (np.integer, int))
```

File: sleap_roots/graphpipeline.py [hunk 2]
```diff
@@ -275,30 +274,38 @@ def get_traits_value_plant(
...
-                plant=plant, frame=frame, lateral_only=False
-            )
-            pts_all_array = pts_all_array.reshape(
-                (1, pts_all_array.shape[0], pts_all_array.shape[1])
-            )
-            pts_all_list = get_all_pts(plant=plant, frame=frame, lateral_only=False)
+        gt_instances_pr = primary.user_instances + primary.unused_predictions
+        gt_instances_lr = lateral.user_instances + lateral.unused_predictions
+
+        # lateral_pts = lateral.numpy()
```
Comment (talmo) [inline sleap_roots/graphpipeline.py:292]: remove unused comment

File: sleap_roots/graphpipeline.py [hunk 3]
```diff
@@ -382,79 +386,74 @@ def get_traits_value_plant_summary(
...
     # get summarized scalar traits per plant
-    data_plant_summary = pd.DataFrame([{"plant_name": data_plant["plant_name"][0]}])
-    column_names = data_plant.columns.tolist()
+    column_names = data_plant_df.columns.tolist()
+    data_plant_frame_summary = {}
     for i in range(len(SCALAR_TRAITS)):
         if SCALAR_TRAITS[i] in column_names:
-            trait = data_plant[SCALAR_TRAITS[i]].values
+            trait = data_plant_df[SCALAR_TRAITS[i]]
+            # LW start
```
Comment (talmo) [inline sleap_roots/graphpipeline.py:445]: remove unused comment

File: sleap_roots/graphpipeline.py [hunk 4]
```diff
@@ -382,79 +386,74 @@ def get_traits_value_plant_summary(
...
+            if trait.shape[0] > 0:
+                if not (
+                    type(trait[0]) == int
+                    or type(trait[0]) == np.int64
+                    or type(trait[0]) == np.float64
+                    or type(trait[0]) == float
```
Comment (talmo) [inline sleap_roots/graphpipeline.py:451]:
```suggestion
                    isinstance(trait[0], (np.floating, float)) 
                    or isinstance(trait[0], (np.integer, int))
```

File: sleap_roots/graphpipeline.py [hunk 5]
```diff
@@ -382,79 +386,74 @@ def get_traits_value_plant_summary(
...
+                ):
+                    values = np.array([element[0] for element in trait])
```
Comment (talmo) [inline sleap_roots/graphpipeline.py:453]: Won't this break if `trait` contains scalars? Like if `type(trait[0]) == int`, then for example:

```
trait = [0, 1, 2]
assert type(trait[0]) == int
element = trait[0]
element[0]  # error
```

File: sleap_roots/graphpipeline.py [hunk 6]
```diff
@@ -490,31 +490,33 @@ def get_traits_value_plant_summary(
...
-            data_plant_frame_summary.columns[j + 2] + "_prc75"
+        data_plant_frame_summary[
+            data_plant_frame_summary_key[j] + "_prc75"
         ] = trait_prc75
-        data_plant_summary[
-            data_plant_frame_summary.columns[j + 2] + "_prc95"
+        data_plant_frame_summary[
+            data_plant_frame_summary_key[j] + "_prc95"
         ] = trait_prc95
+    data_plant_frame_summary["plant_name"] = [os.path.splitext(h5)[0]]
```
Comment (talmo) [inline sleap_roots/graphpipeline.py:512]: prefer `pathlib` over `os.path`

File: sleap_roots/scanline.py [hunk 7]
```diff
@@ -37,62 +36,34 @@ def get_scanline_intersections(
...
 
     intersection = []
-    for i in range(n_line):
-        y_loc = n_interval * (i + 1)
-        line = LineString([(0, y_loc), (width, y_loc)])
 
-        intersection_line = []
-        for j in range(len(points)):
+    for i in range(n_line):  # n_line
+        # y_loc = n_interval * (i + 1)
```
Comment (talmo) [inline sleap_roots/scanline.py:41]: remove unused comment

File: sleap_roots/scanline.py [hunk 8]
```diff
@@ -37,62 +36,34 @@ def get_scanline_intersections(
...
+        horizontal_line_y = n_interval * (i + 1)
+        # line = LineString([(0, y_loc), (width, y_loc)])
```
Comment (talmo) [inline sleap_roots/scanline.py:43]: remove unused comment

File: sleap_roots/scanline.py [hunk 9]
```diff
@@ -37,62 +36,34 @@ def get_scanline_intersections(
...
+
+        intersection_line = 0
+
+        for j in range(len(points) - 1):
+            intersection_counts_root = 0
             # filter out nan nodes
             pts_j = np.array(points[j])[~np.isnan(points[j]).any(axis=1)]
+            # print(pts_j)
```
Comment (talmo) [inline sleap_roots/scanline.py:51]: remove unused comment

File: sleap_roots/scanline.py [hunk 10]
```diff
@@ -37,62 +36,34 @@ def get_scanline_intersections(
...
-                        for point in intersection_root.geoms:
-                            intersection_line.append([point.x, point.y])
+                for k in range(len(pts_j) - 1):
+                    x1, y1 = pts_j[k]
+                    x2, y2 = pts_j[k + 1]
+                    if (y1 >= horizontal_line_y and y2 < horizontal_line_y) or (
+                        y1 < horizontal_line_y and y2 >= horizontal_line_y
+                    ):
+                        current_root += 1
+                        # is_inside_root = not is_inside_root
```
Comment (talmo) [inline sleap_roots/scanline.py:61]: remove unused comment

File: sleap_roots/scanline.py [hunk 11]
```diff
@@ -37,62 +36,34 @@ def get_scanline_intersections(
...
+                intersection_counts_root += current_root
+                # print(intersection_counts_root)
```
Comment (talmo) [inline sleap_roots/scanline.py:63]: remove unused comment

**Review by talmo — CHANGES_REQUESTED** [review]
mostly minor changes, except one special case that may lead to errors

File: sleap_roots/graphpipeline.py (hunk 5 above)
Comment (linwang9926) [reply sleap_roots/graphpipeline.py:453]: I remembered that the trait is like: [[0,1,2]].
The test of the pipeline is good, but I met a problem using Spyder to run code with new environment, will fix it soon

Comment (linwang9926) [reply sleap_roots/graphpipeline.py:446]: I will modify it together with issue #33

**Review by talmo — APPROVED** [review]

**Pull request merged**